### PR TITLE
Refactor `test_python_frontend.py`

### PR DESCRIPTION
**Pull request by Priya2698**
`test_python_frontend.py` is now 4k+ lines which is tedious to work with.

This PR:
1. Moves `exec_nvfuser` into a separate `class NVFuserTest (TestCase)` which can be used as the base class of any nvfuser test class that needs to verify serialization/deserialization.
2. Removes `setUpModule` into `setup_class` of `NVFuserTest` to avoid adding it in every file. This method will run `atexit_serde_check` to enable default serialization at the end of the test suite.
~3. As a starting point, moving matmul and linear tests into a separate file.~ To be done in separate PR.
4. Replaces `unittest` with `pytest` for consistency.

### CODE REVIEW

**Comment by kevinstephano** [comment]
I will note, that if you are just testing a op that you want to use the `pytest_ops.py` infrastructure that also includes checking errors.

**Comment by Priya2698** [comment]
> I will note, that if you are just testing a op that you might want to use the `pytest_ops.py` infrastructure that also includes checking errors.

Right. But it does not test for serialization/deserialization of the new operators.

**Comment by rdspring1** [comment]
@Priya2698 You can take the [serde_check](https://github.com/NVIDIA/Fuser/blob/main/tests/python/test_python_frontend.py#L87-L142) function and apply it as a decorator to [correctness_test_fn](https://github.com/NVIDIA/Fuser/blob/main/tests/python/pytest_ops.py#L135-L154).

You should be able to test serialization on operators like this.

**Comment by rdspring1** [comment]
I created #2753 to test serialization with the pytest framework.

**Comment by Priya2698** [comment]
> I created #2753 to test serialization with the pytest framework.

Thanks, I'll let your merge your PRs and revise this PR accordingly.
On a related note, what is the use of `setUpModule`?

**Comment by rdspring1** [comment]
`setUpModule` tests the automatic serialization upon program exit feature and tests deserializing the default workspace upon creating the tests.

**Comment by Priya2698** [comment]
> `setUpModule` tests the automatic serialization upon program exit feature and tests deserializing the default workspace upon creating the tests.
Is it invoked by `run_tests`? 
I am trying to see how and where it is invoked so I can understand if it needs to be present in all test files, or it can be extracted into utilities.

**Comment by rdspring1** [comment]
`setUpModule` isn't invoked by the programmer. It is done automatically by the `unittest` package.

https://docs.python.org/3/library/unittest.html#setupmodule-and-teardownmodule
https://dmorgan.info/posts/python-unittest-fixtures/

**Comment by Priya2698** [comment]
!build

File: tests/python/test_python_frontend.py [hunk 1]
```diff
@@ -4577,13 +4358,9 @@ def fusion_func(
...
+                with torch.random.fork_rng(devices=[torch.cuda.current_device()]):
+                    with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
+                        ref_out = F.scaled_dot_product_attention(
+                            *qkv, dropout_p=dropout_p, is_causal=is_causal, scale=scale
+                        )
                 torch.testing.assert_close(nvf_out[0], ref_out)
-
-
-if __name__ == "__main__":
-    run_tests()
```
Comment (Priya2698) [inline tests/python/test_python_frontend.py:4589]: Invoking tests through `pytest` does not need this.
@xwang233, in the CI do we invoke tests using `python` anywhere instead of `pytest`?

File: tests/python/test_python_frontend.py [hunk 2]
```diff
@@ -4577,13 +4358,9 @@ def fusion_func(
                 dropout_p = 0.0 if dropout_p is None else dropout_p
                 is_causal = False if is_causal is None else is_causal
 
-                with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
-                    torch.manual_seed(0)
-                    ref_out = F.scaled_dot_product_attention(
-                        *qkv, dropout_p=dropout_p, is_causal=is_causal, scale=scale
-                    )
+                with torch.random.fork_rng(devices=[torch.cuda.current_device()]):
```
Comment (wujingyue) [inline tests/python/test_python_frontend.py:4361]: Would you mind putting the fork_rng changes into a separate PR?

File: tests/python/test_matmul.py
Comment (wujingyue) [inline tests/python/test_matmul.py:1]: Same here -- can this be done in a separate PR? That would make the PR smaller and easier to roll back if needed.

File: tests/python/test_matmul.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
```
Comment (wujingyue) [inline tests/python/test_matmul.py:1]:
```suggestion
# SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
```

File: tests/python/test_python_frontend.py (hunk 2 above)
Comment (Priya2698) [reply tests/python/test_python_frontend.py]: Removed.

File: tests/python/test_matmul.py
Comment (Priya2698) [reply tests/python/test_matmul.py:1]: Restored the tests to `test_python_frontend.py`.

**Comment by Priya2698** [comment]
!build

File: tests/python/test_python_frontend.py (hunk 1 above)
Comment (xwang233) [reply tests/python/test_python_frontend.py:4589]: we're using `python` now, but we can switch to `pytest` if that fits better?

Comment (Priya2698) [reply tests/python/test_python_frontend.py:4589]: Oh interesting. What about `pytest_ops.py` ?
@kevinstephano @rdspring1 Do you have any concerns about switching to `pytest.`

File: tests/python/test_python_frontend.py [hunk 4]
```diff
@@ -3,25 +3,21 @@
...
 import torch.nn.functional as F
-from torch.testing._internal.common_utils import run_tests, TEST_WITH_ROCM, TestCase
-from torch.testing._internal.jit_utils import RUN_CUDA
 import torch._refs as refs
 import torch._prims as prims
 
 from nvfuser import (
-    FusionCache,
     FusionDefinition,
+    FusionCache,
```
Comment (wujingyue) [inline tests/python/test_python_frontend.py:20]: Why changing the order? That being said, it may make more sense to sort alphabetically. (If you use Vim, there's a thing called select and `:sort`.)

**Review by wujingyue — APPROVED** [review]
I'll let @rdspring1 double check the serde changes.

File: tests/python/utils.py [hunk 5]
```diff
@@ -126,3 +128,84 @@ def atexit_serde_check():
     fc = FusionCache.get()
     # Clear FusionCache because the tests expect a new fusion to be generated.
     FusionCache.reset()
+
+
+def serde_check(test_fn: Callable):
```
Comment (rdspring1) [inline tests/python/utils.py:133]: Rename to something like `nvfusertest_serde_check`. This function has specific handling for `NVFuserTest.exec_nvfuser`

**Review by rdspring1 — APPROVED** [review]
LGTM. Just double check that a binary exists in `/tmp/nvf_nvfuser_kernel_db` so that the `atexit` setup is working correctly.

```bash
ls /tmp/nvfuser_kernel_db/
>>> nvf_serde_device8_6_cuda12_4
```

**Comment by rdspring1** [comment]
If we're going to run all tests through `pytest`, we should rename `tests/python/Pytest.md` to `tests/python/README.md`. So, Github will display the markdown automatically.

File: tests/python/test_python_frontend.py (hunk 1 above)
Comment (xwang233) [reply tests/python/test_python_frontend.py:4589]: `pytest` is actually `python -m pytest`, so you can do
```
(gdb) r -m pytest ....py
```

**Comment by Priya2698** [comment]
> If we're going to run all tests through `pytest`, we should rename `tests/python/Pytest.md` to `tests/python/README.md`. So, Github will display the markdown automatically.

I'll update the README in a separate PR.

File: tests/python/test_python_frontend.py (hunk 4 above)
Comment (Priya2698) [reply tests/python/test_python_frontend.py:20]: This was unintentional, possibly crept in while moving things around.

**Comment by rdspring1** [comment]
!build --serde

**Comment by Priya2698** [comment]
!build --serde

**Comment by Priya2698** [comment]
!build --serde

**Comment by Priya2698** [comment]
@xwang233, is the PR safe to merge? I believe the failing tests are unrelated.

**Comment by xwang233** [comment]
> @xwang233, is the PR safe to merge? I believe the failing tests are unrelated.

Is switching from `python` to `pytest` the only change needed in CI? If so, I believe it's ready to merge.

**Comment by Priya2698** [comment]
> > @xwang233, is the PR safe to merge? I believe the failing tests are unrelated.
> 
> Is switching from `python` to `pytest` the only change needed in CI? If so, I believe it's ready to merge.

Yes, we should invoke tests using `pytest`.
We can change that in the CI first and re-run the python test suite to be safe.

**Comment by xwang233** [comment]
> > > @xwang233, is the PR safe to merge? I believe the failing tests are unrelated.
> > 
> > 
> > Is switching from `python` to `pytest` the only change needed in CI? If so, I believe it's ready to merge.
> 
> Yes, we should invoke tests using `pytest`. We can change that in the CI first and re-run the python test suite to be safe.

Thanks. I just changed that for Github CI. Feel free to start a new pipeline.

**Comment by Priya2698** [comment]
!build --serde
